server/routerlicious/packages/services-shared/src/socketIoServer.ts

```diff
@@ -127,6 +132,64 @@ class SocketIoServer implements core.IWebSocketServer {
 	}
 
 	public async close(): Promise<void> {
+		const sleep = async (timeMs: number) =>
+			new Promise((resolve) => setTimeout(resolve, timeMs));
+
+		if (this.socketIoConfig?.gracefulShutdownEnabled) {
+			// Gradual disconnection of websocket connections
+			const drainTime = this.socketIoConfig?.gracefulShutdownDrainTimeMs ?? 30000;
+			const drainInterval = this.socketIoConfig?.gracefulShutdownDrainIntervalMs ?? 1000;
+			if (drainTime > 0 && drainInterval > 0) {
+				// we are assuming no new connections appear once we start. any leftover connections will be closed when close is called
+				const connections = await this.io.fetchSockets();
+				const connectionCount = connections.length;
+				const telemetryProperties = {
+					drainTime,
+					drainInterval,
+					connectionCount,
+				};
+				Lumberjack.info("Graceful disconnection started", telemetryProperties);
+				const metricForTimeTaken = Lumberjack.newLumberMetric(
+					LumberEventName.GracefulShutdown,
+					telemetryProperties,
+				);
+				// total number of drains to run
+				const totalDrains = Math.ceil(drainTime / drainInterval);
+				// number of connections to disconnect per drain
+				const connectionsToDisconnectPerDrain = Math.ceil(connectionCount / totalDrains);
+				let done = false;
+				const drainConnections = Array.from(connections.values());
+				if (connectionsToDisconnectPerDrain > 0) {
+					// start draining                let done = false;
+					let n = 0;
+					for (let i = 0; i < totalDrains; i++) {
+						for (let j = 0; j < connectionsToDisconnectPerDrain; j++) {
+							const connection: RemoteSocket<any, any> = drainConnections[n];
+							if (!connection) {
+								done = true;
+								break;
+							}
+							try {
+								connection.disconnect(true);
+							} catch (e) {
+								Lumberjack.error("Graceful disconnect exception", undefined, e);
```

Comment:
Log metricForTimeTaken.error() as well here?

Reply:
on second though, you're not closing disconnect on catch here. So, calling error() on metric won't work.
We need a better way to report errors for this metric?

Reply:
We can alert on the duration exceeding some number or being too short. I agree there is no good way, also, most of this flow is already captured by the runnerMetric with  "encountered an error during stop" logs